File: geocode-v1/src/main/java/org/microg/nlp/geocode/v1/GeocodeProvider.java
/*
 * SPDX-FileCopyrightText: 2019, microG Project Team
 * SPDX-License-Identifier: Apache-2.0
 */

package org.microg.nlp.geocode.v1;

import android.content.Context;
import android.location.Address;
import android.location.GeocoderParams;
import android.util.Log;

import org.microg.nlp.client.UnifiedLocationClient;

import java.util.List;

public class GeocodeProvider extends com.android.location.provider.GeocodeProvider {
    private static final String TAG = "UGeocode";
    private Context context;
    private static final long TIMEOUT = 10000;

    public GeocodeProvider(Context context) {
        this.context = context;
        UnifiedLocationClient.get(context).ref();
    }

    public void onDisable() {
        UnifiedLocationClient.get(context).unref();
    }

    @Override
    public String onGetFromLocation(double latitude, double longitude, int maxResults, GeocoderParams params, List<Address> addrs) {
        try {
            return handleResult(addrs, UnifiedLocationClient.get(context).getFromLocationSync(latitude, longitude, maxResults, params.getLocale().toString(), TIMEOUT));
        } catch (Exception e) {
            Log.w(TAG, e);
            return e.getMessage();
        }
    }

    @Override
    public String onGetFromLocationName(String locationName, double lowerLeftLatitude, double lowerLeftLongitude, double upperRightLatitude, double upperRightLongitude, int maxResults, GeocoderParams params, List<Address> addrs) {
        try {
            return handleResult(addrs, UnifiedLocationClient.get(context).getFromLocationNameSync(locationName, maxResults, lowerLeftLatitude, lowerLeftLongitude, upperRightLatitude, upperRightLongitude, params.getLocale().toString(), TIMEOUT));
        } catch (Exception e) {
            Log.w(TAG, e);
            return e.getMessage();
        }
    }

    private String handleResult(List<Address> realResult, List<Address> fuserResult) {
        if (fuserResult.isEmpty()) {
            return "no result";
        } else {
            realResult.addAll(fuserResult);
            return null;
        }
    }
}

File: geocode-v1/src/main/java/org/microg/nlp/geocode/v1/GeocodeProvider.kt
/*
 * SPDX-FileCopyrightText: 2019, microG Project Team
 * SPDX-License-Identifier: Apache-2.0
 */
package org.microg.nlp.geocode.v1

import android.content.Context
import android.location.Address
import android.location.GeocoderParams
import android.os.Bundle
import android.util.Log
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import com.android.location.provider.GeocodeProvider
import org.microg.nlp.client.GeocodeClient
import org.microg.nlp.service.api.GeocodeRequest
import org.microg.nlp.service.api.LatLon
import org.microg.nlp.service.api.LatLonBounds
import org.microg.nlp.service.api.ReverseGeocodeRequest

class GeocodeProvider(context: Context, lifecycle: Lifecycle) : GeocodeProvider() {
    private val client: GeocodeClient = GeocodeClient(context, lifecycle)

    override fun onGetFromLocation(latitude: Double, longitude: Double, maxResults: Int, params: GeocoderParams, addrs: MutableList<Address>): String? {
        return try {
            handleResult(addrs, client.requestReverseGeocodeSync(
                    ReverseGeocodeRequest(LatLon(latitude, longitude), maxResults, params.locale),
                    Bundle().apply { putString("packageName", params.clientPackage) }
            ))
        } catch (e: Exception) {
            Log.w(TAG, e)
            e.message
        }
    }

    override fun onGetFromLocationName(locationName: String?, lowerLeftLatitude: Double, lowerLeftLongitude: Double, upperRightLatitude: Double, upperRightLongitude: Double, maxResults: Int, params: GeocoderParams, addrs: MutableList<Address>): String? {
        return try {
            handleResult(addrs, client.requestGeocodeSync(
                    GeocodeRequest(locationName!!, LatLonBounds(LatLon(lowerLeftLatitude, lowerLeftLongitude), LatLon(upperRightLatitude, upperRightLongitude)), maxResults, params.locale),
                    Bundle().apply { putString("packageName", params.clientPackage) }
            ))
        } catch (e: Exception) {
            Log.w(TAG, e)
            e.message
        }
    }

    private fun handleResult(realResult: MutableList<Address>, fuserResult: List<Address>): String? {
        return if (fuserResult.isEmpty()) {
            "no result"
        } else {
            realResult.addAll(fuserResult)
            null
        }
    }

    suspend fun connect() = client.connect()
    suspend fun disconnect() = client.disconnect()

    companion object {
        private const val TAG = "GeocodeProvider"
        private const val TIMEOUT: Long = 10000
    }
}
